R/aggregate.R

```diff
@@ -0,0 +1,180 @@
+##' @title Return the Robust Expression Summary of a matrix
+##'
+##' @description
+##' This function calculates the robust summarisation for each feature
+##' (protein). Note that the function assumes that the intensities in
+##' input `e` are already log-transformed.
+##'
+##' @param x A feature by sample `matrix` containing quantitative
+##'     data with mandatory `colnames` and `rownames`.
+##' @param ... Additional arguments passed to [MASS::rlm()].
+##' @return `numeric()` vector of length `ncol(x)` with robust
+##'     summarised values.
+##' 
+##' @author Adriaan Sticker, Sebastian Gibb and Laurent Gatto
+##' 
+##' @export
+##' 
+##' @examples
+##' x <- matrix(rnorm(30), nrow = 3)
+##' colnames(x) <- letters[1:10]
+##' rownames(x) <- LETTERS[1:3]
+##' robustSummary(x)
+robustSummary <- function(x, ...) {
+    if (is.null(colnames(x)))
+        stop("colnames must not be empty.")
+    if (is.null(rownames(x)))
+        stop("rownames must not be empty.")
+
+    ## If there is only one 1 peptide for all samples return
+    ## expression of that peptide
+    if (nrow(x) == 1L) return(x)
+
+    ## remove missing values
+    p <- !is.na(x)
+    expression <- x[p] ## expression becomes a vector
+    sample <- rep(colnames(x), each = nrow(x))[p]
+    feature <- rep(rownames(x), times = ncol(x))[p]
+
+    ## model.matrix breaks on factors with 1 level so make vector of
+    ## ones (intercept).
+    if (length(unique(sample)) == 1L) sample <- rep(1, length(sample))
+
+    ## Sum contrast on peptide level so sample effect will be mean
+    ## over all peptides instead of reference level.
+    X <- stats::model.matrix(~ -1 + sample + feature,
```

Comment:
Is there any reason to not import this function but to call it with `stats::`? Note that a call using `::` is slower than one without (according to Hadleys book - can't remember exactly where).

Reply:
No specific reason, just to avoid confusion - I wouldn't be surprised of there are other `model.matrix` functions around. 

In terms of speed, I honestly don't think it matters at all - I'm pretty sure this difference is insignificant in the wider context of aggregation.

Reply:
Then I would add an `@importFrom stats model.matrix` to import it - you can then also be sure that that version/implementation is used.